lms/services/canvas_api.py

```diff
@@ -33,202 +48,351 @@ def get_token(self, authorization_code):
+        :raise CanvasAPIServerError: if we do have an access token but the
+            Canvas API request fails for any other reason
 
         :return: a list of raw section dicts as received from the Canvas API
         :rtype: list(dict)
         """
-        oauth2_token = self._oauth2_token
-
-        return self.send_with_refresh_and_retry(
-            self._helper.authenticated_users_sections_request(
-                oauth2_token.access_token, course_id
-            ),
-            CanvasAuthenticatedUsersSectionsResponseSchema,
-            oauth2_token.refresh_token,
+
+        # Canvas's sections API
+        # (https://canvas.instructure.com/doc/api/sections.html) only allows
+        # you to get _all_ of a course's sections, it doesn't provide a way to
+        # get only the sections that the authenticated user belongs to. So we
+        # have to get the authenticated user's sections from part of the
+        # response from a courses API endpoint instead.
+        #
+        # Canvas's "Get a single course" API is capable of doing this if the
+        # ?include[]=sections query param is given:
+        #
+        #    https://canvas.instructure.com/doc/api/courses.html#method.courses.show
+        #
+        # The ?include[]=sections query param is documented elsewhere (in the
+        # "List your courses" API:
+        # https://canvas.instructure.com/doc/api/courses.html#method.courses.index)
+        # as:
+        #
+        #    "Section enrollment information to include with each Course.
+        #    Returns an array of hashes containing the section ID (id), section
+        #    name (name), start and end dates (start_at, end_at), as well as the
+        #    enrollment type (enrollment_role, e.g. 'StudentEnrollment')."
+        #
+        # In practice ?include[]=sections seems to add a "sections" key to the
+        # API response that is a list of section dicts, one for each section
+        # the authenticated user is currently enrolled in, each with the
+        # section's "id" and "name" among other fields.
+        #
+        # **We don't know what happens if the user belongs to a really large
+        # number of sections**. Does the list of sections embedded within the
+        # get course API response just get really long? Does it get truncated?
+        # Can you paginate through it somehow? This seems edge-casey enough
+        # that we're ignoring it for now.
+
+        return self._make_authenticated_request(
+            "GET",
+            f"courses/{course_id}",
+            params={"include[]": "sections"},
+            schema=self._AuthenticatedUsersSectionsSchema,
+        )
+
+    class _AuthenticatedUsersSectionsSchema(RequestsResponseSchema):
+        """Schema for the "authenticated user's sections" responses."""
+
+        sections = fields.List(
+            fields.Nested(_SectionSchema),
+            validate=validate.Length(min=1),
+            required=True,
```

Comment:
A good demonstration here of the:

 * Here's how we make the call
 * Here's what we expect back